### ISSUE TITLE
renderdoc.dll calls WSAStartup/ WSACleanup from DLLMain

### ISSUE THREAD

**Issue by RanXacT** [baldurk/renderdoc#2079]
## Description

I've been seeing rare deadlocks on test machines for some time now, and tracked this down to deadlocks within WSACleanup on shutdown. Internally WSACleanup is stuck on a lock and waits forever. WSACleanup is not allowed to be called from DLLMain, and deadlock is listed as a potential outcome of doing this. 

From the documentation:
```
the WSACleanup function should not be called from the DllMain function in a application DLL. This can potentially cause deadlocks.
```
https://docs.microsoft.com/en-us/windows/win32/api/winsock/nf-winsock-wsacleanup

I haven't expclicitly caught a deadlock within WSAStartup, but it is also called within DLLMain. and it also contains the same warning in the documentaiton: 
https://docs.microsoft.com/en-us/windows/win32/api/winsock/nf-winsock-wsastartup

I have locally created explicit Init/Destroy functions in the dll API to move these calls out of DLLMain, but this would break API compatibility for renderdoc so I doubt you'd want this as a PR. It seems like a discussion about whether you want to fix this bug at all, and or how you would like to fix it should occur first.

## Steps to reproduce

* Set a breakpoint on `Network::Init` and `Network::Shutdown`. Specifically on calls to `WSAStartup` and `WSACleanup`
* Launch any application
* Load renderdoc.dll
* Observe `DLLMain` on callstack.

Callstack for `Network::Init`:
```
>	renderdoc.dll!Network::Init() Line 107	C++
 	renderdoc.dll!RenderDoc::Initialise() Line 350	C++
 	renderdoc.dll!add_hooks() Line 66	C++
 	renderdoc.dll!DllMain(HINSTANCE__ * hModule, unsigned long ul_reason_for_call, void * lpReserved) Line 77	C++
 	renderdoc.dll!dllmain_dispatch(HINSTANCE__ * const instance, const unsigned long reason, void * const reserved) Line 195	C++
 	ntdll.dll!LdrpCallInitRoutine()	Unknown
 	ntdll.dll!LdrpInitializeNode()	Unknown
 	ntdll.dll!LdrpInitializeGraphRecurse()	Unknown
 	ntdll.dll!LdrpPrepareModuleForExecution()	Unknown
 	ntdll.dll!LdrpLoadDllInternal()	Unknown
 	ntdll.dll!LdrpLoadDll()	Unknown
 	ntdll.dll!LdrLoadDll()	Unknown
 	KernelBase.dll!LoadLibraryExW()	Unknown
 	KernelBase.dll!LoadLibraryExA()	Unknown
 	KernelBase.dll!LoadLibraryA()	Unknown
```

Callstack for `Network::Shutdown`:
```
>	renderdoc.dll!Network::Shutdown() Line 112	C++
 	renderdoc.dll!RenderDoc::~RenderDoc() Line 513	C++
 	renderdoc.dll!`RenderDoc::Inst'::`2'::`dynamic atexit destructor for 'realInst''()	C++
 	ucrtbase.dll!<lambda>(void)()	Unknown
 	ucrtbase.dll!__crt_seh_guarded_call<int>::operator()<<lambda_7777bce6b2f8c936911f934f8298dc43>,<lambda>(void) &,<lambda_3883c3dff614d5e0c5f61bb1ac94921c>>()	Unknown
 	ucrtbase.dll!_execute_onexit_table()	Unknown
 	renderdoc.dll!dllmain_crt_process_detach(const bool is_terminating) Line 109	C++
 	renderdoc.dll!dllmain_dispatch(HINSTANCE__ * const instance, const unsigned long reason, void * const reserved) Line 207	C++
 	ntdll.dll!LdrpCallInitRoutine()	Unknown
 	ntdll.dll!LdrShutdownProcess()	Unknown
 	ntdll.dll!RtlExitUserProcess()	Unknown
 	kernel32.dll!ExitProcessImplementation()	Unknown
 ```
Note that static destructors are called from the c library dllmain_dispatch (the function that calls the client DLLMain) and this is for all intents and purposes still DLLMain. 

## Environment

* RenderDoc version: 1.10 (although also fails on earlier versions, I've been seeing the issue for quite some time, but hadn't caught it in a debugger until just recently).
* Operating System: Windows 10
* Graphics API: d3d12 (doesn't matter)

**Comment by baldurk** [comment]
Yeh unfortunately on paper DllMain is pretty crippled and there's almost nothing interesting that's safe to do inside it. We skate around the rules and do as little as we can get away with because there's nowhere else to do initialisation.

The only solution I can see is to just not call `WSACleanup` and let it leak, since that's where you've seen the deadlocks happen. I know it's still not safe to call `WSAStartup` either but I don't see an alternative. We only shutdown on process shutdown so it will all be cleaned up anyway.

**Comment by RanXacT** [comment]
I suspected as much. In practice I agree that it should be safe to skip the call to WSACleanup. A correctness conundrum, because we like to fulfill our API requirements :). but it does seem it's the lesser evil. That said I guess people who don't run many thousands of tests a day would be ok with rare deadlocks...

Since I'm in a fork the explicit init / shutdown calls are working ok for me and give me a place to move more code from dllmain should I find more. So I'll probably keep with this for now on my end.

**Comment by baldurk** [comment]
Yeh if there were any way I could avoid this I would, as I don't like relying on behaviour which as good reason to be undefined/non-contractual. Unfortunately when it comes to injecting code into processes understandably this is not exactly a blessed path :).

For your own use then certainly you can have explicit initialise and shutdown calls, but that's sadly not reasonable to ask everyone to do to be able to use renderdoc.

**Comment by RanXacT** [comment]
An optional explicit shutdown for network seems like a viable option. Something along the lines of:

**Comment by RanXacT** [comment]
An optional explicit shutdown for network seems like a viable option. Something along the lines of:

```
// Optional function to shutdown RenderDoc's network communication.
//
// Note: this will disable any future communication with a replay app.
// and may crash if called while interacting with a replay app.
// Calling this before shutdown can avoid rare deadlock condition
// on windows upon process exit. 
typedef void(RENDERDOC_CC *pRENDERDOC_ShutdownNetwork)();
```

Along Something like:
```
void RenderDoc::ShutdownNework()
{
  if(m_RemoteThread)
  {
    // explicitly wait for thread to shutdown, this call is not from module unloading and
    // we want to be sure everything is gone before we remove our module & hooks
    m_TargetControlThreadShutdown = true;
    Threading::JoinThread(m_RemoteThread);
    Threading::CloseThread(m_RemoteThread);
    m_RemoteThread = 0;
  }

  if(m_NetworkInitialized)
  {
  	Network::Shutdown();
  	m_NetworkInitialized = false;
  }
}
```

And then protect `Network::Shutdown()` in `RenderDoc::~RenderDoc` behind m_NetworkInitialized.

As a general plan that seems reasonable, and allows compatibility with existing apps (and apps which choose not to call the function).

**Comment by RanXacT** [comment]
that said I'm ok with any outcome here LMK. i can potentially make changes if we agree on a solution. 

Options as I see them: 
* Do nothing. In this case I suggest adding comments on these calls to make it clear the rule breaking is intentional, and what kinds of issues can result (deadlock). but that such issues are rare and hard to avoid. There are similar comments on for example the remote thread join (or lack thereof), etc.
* Remove WSACleanup entirely. Leaving a commented call for people who wander open source repos for gotcha's that really aren't useful :) as you say it's really not necessary since we are shutting down in this case.
* Add an optional shutdown network call. as described above.

**Comment by baldurk** [comment]
I think removing WSACleanup is the sensible approach. As we've said 'leaking' does nothing meaningful when we're not continually leaking over time and it's only really an improper shutdown. We do this for other things like global one-off allocs we don't free and forcing driver DLLs to stay loaded to avoid driver bugs so it wouldn't be the first thing.

Adding an optional shutdown call is technically possible but I suspect the number of people who'd actually use it aside from yourself is a rounding error. Adding extra API surface doesn't seem worth it for this - it would really only be if we did nothing because there was a serious enough downside to whatever shutdown fix we had to make such that we decided keeping the chance of deadlock (or whatever else) is preferable. Then an optional shutdown to avoid the deadlock would be worth adding.
